monkey/monkey_island/cc/environment/environment_config.py

```diff
@@ -6,56 +6,48 @@
 from typing import Dict, List
 
 import monkey_island.cc.environment.server_config_generator as server_config_generator
-from monkey_island.cc.consts import MONKEY_ISLAND_ABS_PATH
+from monkey_island.cc.consts import DEFAULT_SERVER_CONFIG_PATH
 from monkey_island.cc.environment.user_creds import UserCreds
 from monkey_island.cc.resources.auth.auth_user import User
 from monkey_island.cc.resources.auth.user_store import UserStore
 
-SERVER_CONFIG_FILENAME = "server_config.json"
-
 
 class EnvironmentConfig:
-    def __init__(self,
-                 server_config: str,
-                 deployment: str,
-                 user_creds: UserCreds,
-                 aws=None):
-        self.server_config = server_config
-        self.deployment = deployment
-        self.user_creds = user_creds
-        self.aws = aws
-
-    @staticmethod
-    def get_from_json(config_json: str) -> EnvironmentConfig:
-        data = json.loads(config_json)
-        return EnvironmentConfig.get_from_dict(data)
+    def __init__(self, file_path):
+        self._server_config_path = os.path.expanduser(file_path)
+        self.server_config = None
+        self.deployment = None
+        self.user_creds = None
+        self.aws = None
 
-    @staticmethod
-    def get_from_dict(dict_data: Dict) -> EnvironmentConfig:
-        user_creds = UserCreds.get_from_dict(dict_data)
-        aws = dict_data['aws'] if 'aws' in dict_data else None
-        return EnvironmentConfig(server_config=dict_data['server_config'],
-                                 deployment=dict_data['deployment'],
-                                 user_creds=user_creds,
-                                 aws=aws)
+        self._load_from_file(self._server_config_path)
 
-    def save_to_file(self):
-        file_path = EnvironmentConfig.get_config_file_path()
-        with open(file_path, 'w') as f:
-            f.write(json.dumps(self.to_dict(), indent=2))
+    def _load_from_file(self, file_path=DEFAULT_SERVER_CONFIG_PATH):
+        file_path = os.path.expanduser(file_path)
 
-    @staticmethod
-    def get_from_file() -> EnvironmentConfig:
-        file_path = EnvironmentConfig.get_config_file_path()
         if not Path(file_path).is_file():
             server_config_generator.create_default_config_file(file_path)
         with open(file_path, 'r') as f:
             config_content = f.read()
-        return EnvironmentConfig.get_from_json(config_content)
 
-    @staticmethod
-    def get_config_file_path() -> str:
-        return os.path.join(MONKEY_ISLAND_ABS_PATH, 'cc', SERVER_CONFIG_FILENAME)
+        self._load_from_json(config_content)
+
+    def _load_from_json(self, config_json: str) -> EnvironmentConfig:
```

Comment:
* Could be pure function - you don't need `self` to load json contents

Reply:
`_load_from_json()` calls `_load_from_dict()` which populates member variables.

Reply:
Same as with `_load_from_file` (I don't know where did that comment go 🤔), the contents of these functions have nothing to do with object itself (except that there is one method in the chain of calls that actually do). Instead, you could have pure functions if you get the contents of config first and set them in the constructor. It's not a requirement, just a step towards more testable code.

Reply:
The issue is that the object needs the path to the config file so it can write to it later. This could also be passed to the constructor, but this interface allows you to create somewhat invalid objects, i.e. nothing enforces that the contents of the file path actually match the properties of the object. By providing the file path to the constructor and constructing the object from that file, you create an interface that helps ensure the file contents and the objects properties are in sync.

One solution to this would be to extract the write-to-file capability from `EnvironmentConfig` and have a different  component handle that, but that level of refactor is out of scope for what I'm trying to achieve.

Reply:
I think I wasn't clear enough or I don't understand you reply. My keypoint is that `_load_from_json` and `_load_from_file` can be refactored into pure functions (static methods). See my suggestions

Reply:
Since they're private I'm not sure what it you get by switching them. Similarly, since they're private, I don't see what it would hurt. Let's talk about it over zoom when we discuss default environments.